### PR TITLE
BUG: Fix using "inf"/"-inf" in na_values for csv with int index column

### LINKED ISSUE #17128: OverflowError in read_csv when specifying certain na_values
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
from pandas.compat import StringIO
data = StringIO("a,b,c\n1,2,3\n4,5,6\n7,8,9")
na_values = ['-inf']
index_col = 0
df = pd.read_csv(data, na_values=na_values, index_col=index_col)
```
#### Problem description

`read_csv()` fails with the following traceback when specifying certain `na_values` with `index_col`:

```
Traceback (most recent call last):
  File "run.py", line 9, in <module>
    df = pd.read_csv(data, na_values=na_values, index_col=index_col)
  File "/home/liauys/Code/pandas/pandas/io/parsers.py", line 660, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "/home/liauys/Code/pandas/pandas/io/parsers.py", line 416, in _read
    data = parser.read(nrows)
  File "/home/liauys/Code/pandas/pandas/io/parsers.py", line 1010, in read
    ret = self._engine.read(nrows)
  File "/home/liauys/Code/pandas/pandas/io/parsers.py", line 1837, in read
    index, names = self._make_index(data, alldata, names)
  File "/home/liauys/Code/pandas/pandas/io/parsers.py", line 1347, in _make_index
    index = self._agg_index(index)
  File "/home/liauys/Code/pandas/pandas/io/parsers.py", line 1440, in _agg_index
    arr, _ = self._infer_types(arr, col_na_values | col_na_fvalues)
  File "/home/liauys/Code/pandas/pandas/io/parsers.py", line 1524, in _infer_types
    mask = algorithms.isin(values, list(na_values))
  File "/home/liauys/Code/pandas/pandas/core/algorithms.py", line 408, in isin
    values, _, _ = _ensure_data(values, dtype=dtype)
  File "/home/liauys/Code/pandas/pandas/core/algorithms.py", line 74, in _ensure_data
    return _ensure_int64(values), 'int64', 'int64'
  File "pandas/_libs/algos_common_helper.pxi", line 3227, in pandas._libs.algos.ensure_int64
  File "pandas/_libs/algos_common_helper.pxi", line 3232, in pandas._libs.algos.ensure_int64
OverflowError: cannot convert float infinity to integer
```

Any of the following makes the error go away:

* The index column does contain the said NA value
* Using `na_values` of `['inf']` instead of `['-inf']`
* Not specifying index_col
* Using version 0.19 or older

#### Expected Output

There should not be any error.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.13.final.0
python-bits: 64
OS: Linux
OS-release: 4.11.9-1-ARCH
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.21.0.dev+316.gf2b0bdc9b
pytest: None
pip: 9.0.1
setuptools: 36.2.5
Cython: 0.26
numpy: 1.13.1
scipy: None
pyarrow: None
xarray: None
IPython: 5.4.1
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by Templarrr**
- [x] closes #17128
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

The issue happens when you try to use 'inf' or '-inf' as part of na_values in read_csv.

Code snippet to reproduce:
```
from StringIO import StringIO

import pandas as pd

dataset = StringIO('''index,col1,col2,col3
1,6,10,14
2,7,11,15
3,8,12,16
4,9,13,17
5,inf,-inf,bla
''')
na_values = ['inf', '-inf', 'bla']

df = pd.read_csv(dataset, na_values=na_values, index_col='index')
print df
print df.dtypes
```

Without fix:
```
Traceback (most recent call last):
  File "/home/modintsov/workspace/DataRobot/playground.py", line 39, in <module>
    df = pd.read_csv(dataset, na_values=na_values, index_col='index')
  File "/home/modintsov/.virtualenvs/dev/local/lib/python2.7/site-packages/pandas/io/parsers.py", line 678, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "/home/modintsov/.virtualenvs/dev/local/lib/python2.7/site-packages/pandas/io/parsers.py", line 446, in _read
    data = parser.read(nrows)
  File "/home/modintsov/.virtualenvs/dev/local/lib/python2.7/site-packages/pandas/io/parsers.py", line 1036, in read
    ret = self._engine.read(nrows)
  File "/home/modintsov/.virtualenvs/dev/local/lib/python2.7/site-packages/pandas/io/parsers.py", line 1922, in read
    index, names = self._make_index(data, alldata, names)
  File "/home/modintsov/.virtualenvs/dev/local/lib/python2.7/site-packages/pandas/io/parsers.py", line 1426, in _make_index
    index = self._agg_index(index)
  File "/home/modintsov/.virtualenvs/dev/local/lib/python2.7/site-packages/pandas/io/parsers.py", line 1520, in _agg_index
    arr, _ = self._infer_types(arr, col_na_values | col_na_fvalues)
  File "/home/modintsov/.virtualenvs/dev/local/lib/python2.7/site-packages/pandas/io/parsers.py", line 1600, in _infer_types
    mask = algorithms.isin(values, list(na_values))
  File "/home/modintsov/.virtualenvs/dev/local/lib/python2.7/site-packages/pandas/core/algorithms.py", line 418, in isin
    values, _, _ = _ensure_data(values, dtype=dtype)
  File "/home/modintsov/.virtualenvs/dev/local/lib/python2.7/site-packages/pandas/core/algorithms.py", line 82, in _ensure_data
    return _ensure_int64(values), 'int64', 'int64'
  File "pandas/_libs/algos_common_helper.pxi", line 3227, in pandas._libs.algos.ensure_int64
  File "pandas/_libs/algos_common_helper.pxi", line 3232, in pandas._libs.algos.ensure_int64
OverflowError: cannot convert float infinity to integer
```

With fix (as expected):
```
       col1  col2  col3
index
1       6.0  10.0  14.0
2       7.0  11.0  15.0
3       8.0  12.0  16.0
4       9.0  13.0  17.0
5       NaN   NaN   NaN
col1    float64
col2    float64
col3    float64
dtype: object
```

### CODE REVIEW

File: doc/source/whatsnew/v0.23.4.txt [hunk 1]
```diff
@@ -41,7 +41,7 @@ Bug Fixes
 
 **Indexing**
 
--
+- Fix OverflowError when trying to use 'inf' as na_value with int index column
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.4.txt]: add the issue number here, use double-backticks around OverflowError. use double backticks on na_value; spell out int -> integer.

File: pandas/tests/io/parser/na_values.py [hunk 2]
```diff
@@ -369,3 +369,13 @@ def test_no_na_filter_on_index(self):
         expected = DataFrame({"a": [1, 4], "c": [3, 6]},
                              index=Index([np.nan, 5.0], name="b"))
         tm.assert_frame_equal(out, expected)
+
+    def test_inf_na_values_with_int_index(self):
+        data = "idx,col1,col2\n1,3,4\n2,inf,-inf"
```
Comment (jreback) [inline pandas/tests/io/parser/na_values.py]: add the issue number here as a comment

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by Templarrr** [comment]
Oh, wow, it's actually my colleague :) 
Didn't know Liau Yung Siang reported that already.

Thanks for the review, it's already late In my timezone, but I'll update the PR tomorrow!

**Comment by Templarrr** [comment]
@jreback  I've confirmed that this solves the error that @YS-L reported in #17128 and addressed your PR review comments. Can you look again?

Also yesterday the 0.23.4 changelog got updated with today's date, but I don't see on pypi newer pandas version released yet, so I'm not entirely sure - will my bugfix be a part of 0.23.4 or should I move the comment to 0.24.0 changelog?

**Comment by Templarrr** [comment]
The failure in travis is some kind of network glitch, unrelated to these changes :(

**Comment by Templarrr** [comment]
@jreback I see 0.23.4 was released, so this fix obviously didn't make it :) I've moved the changelog line to 0.24.0

**Comment by Templarrr** [comment]
@jreback I see you've moved the label to 0.23.5, didn't know there was 0.23.5 planned :) 
There is no 0.23.5 changelog currently, should I create a new one and move my changelog line there?

**Comment by jreback** [comment]
yeah we might do a 0.23.5. if you want to push a 0.23.5 whats (empty) pls do , but a new PR pls.

File: doc/source/whatsnew/v0.24.0.txt [hunk 3]
```diff
@@ -607,6 +607,7 @@ Indexing
 - Fixed ``DataFrame[np.nan]`` when columns are non-unique (:issue:`21428`)
 - Bug when indexing :class:`DatetimeIndex` with nanosecond resolution dates and timezones (:issue:`11679`)
 - Bug where indexing with a Numpy array containing negative values would mutate the indexer (:issue:`21867`)
+- Fix ``OverflowError`` when trying to use 'inf' as ``na_value`` with integer index column (:issue:`17128`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: can you move to 0.23.5 (rebase on master to see the whatsnew)

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.0.txt (hunk 3 above)
Comment (Templarrr) [reply doc/source/whatsnew/v0.24.0.txt]: will do in few minutes

Comment (Templarrr) [reply doc/source/whatsnew/v0.24.0.txt]: @jreback  done.

**Review by jreback — APPROVED** [review]

File: doc/source/whatsnew/v0.23.5.txt [hunk 4]
```diff
@@ -40,3 +40,7 @@ Bug Fixes
 
 -
 -
+
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.5.txt]: can you put this in the IO section, and referencde `:func:read_csv` here

**Review by jreback — CHANGES_REQUESTED** [review]
ping on green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @Templarrr

**Comment by Templarrr** [comment]
You're welcome! Always glad to help :)

**Pull request merged**